File: General/25-reverse-nodes-in-k-group/main.cpp
#include <vector>
using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};

class Solution {
   public:
    pair<ListNode*, ListNode*> _reverseSection(ListNode* sec_start, ListNode* sec_end) {
        ListNode *prev = nullptr, *curr = sec_start, *next = nullptr;
        while (curr != sec_end) {
            next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }

        // section endpoints (new_head, new_tail)
        return {prev, sec_start};
    }

    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* dummy = new ListNode(0);
        dummy->next = head;

        ListNode* sec_start = head;
        ListNode* sec_end = sec_start;
        ListNode* prev = dummy;

        while (sec_start) {
            for (int i = 0; i < k; ++i) {
                if (!sec_end) {
                    return dummy->next;
                }
                sec_end = sec_end->next;
            }
            auto sec_endpoints = _reverseSection(sec_start, sec_end);
            prev->next = sec_endpoints.first;
            sec_endpoints.second->next = sec_end;

            sec_start = sec_end;
            prev = sec_endpoints.second;
        }

        return dummy->next;
    }
};

File: General/25-reverse-nodes-in-k-group/main.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def _reverse_segment(self, begin: ListNode, end: ListNode):
        prev = None
        curr = begin
        while curr is not None and curr != end:
            # prev .. curr -> next
            next = curr.next
            curr.next = prev
            prev = curr
            curr = next

        return (prev, begin)

    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(-1, head)
        slow = head
        fast = head

        prev = dummy
        while fast is not None:

            for _ in range(k):
                if fast is None:
                    return dummy.next
                fast = fast.next

            new_begin, new_end = self._reverse_segment(slow, fast)
            prev.next = new_begin
            new_end.next = fast

            prev = new_end
            slow = fast

        return dummy.next
